### ISSUE TITLE
Clean Shutdown Proposal

### ISSUE THREAD

**Issue by DavidVorick** [NebulousLabs/writeaheadlog#4]
This issue has a minor dependency on the metadata issue, which means that if two people are working on them in parallel, there will be small merge conflicts. The merge conflicts will be minor enough that work can begin on this issue immediately, there is no need to wait until the metadata pull request is merged.

When the wal is opened, there should a field on disk (stored in the metadata) indicating whether the previous shutdown was clean or unclean. When you load metadata, the function will return either a value indicating clean shutdown, or a value indicating unclean shutdown.

If clean shutdown is indicated, the wal should skip recovery entirely, write new metadata indicating unclean shutdown (need to write out the full 4096 bytes to obliterate any recovery data that exists on disk - that's to be discussed in another issue).

If unclean shutdown is indicated, the the `recover` function should be called. When recovery completes (recover is not complete until the user calls `RecoveryComplete()`), write new metadata indicating unclean shutdown (need to write out the full 4096 bytes to obliterate any recovery data that exists on disk - that's to be discussed in another issue).

When the wal shutds down, a check needs to be made that there are no outstanding transactions. A transaction cannot be cleared until the creator calls `SignalUpdatesApplied`. If there are transactions that did not call this, we cannot write the shutdown as clean, and instead write the shutdown as unclean. Just like at startup, we'll also write out the full 4096 to obliterate any recovery data.

If all transactions have called `SignalUpdatesApplied`, then the wal has cleaned up fully and can write the metadata header to indicate clean shutdown. Just like with the other writes, we will need to write the full 4096 byte page to obliterate the recovery data.

The purpose of the recovery data itself will be covered in another issue.
